File: 101-200/101.symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True
        
        return self.isTreeSymmetric(root.left, root.right)

    def isTreeSymmetric(self, p: TreeNode, q: TreeNode) -> bool:
        if p and q:
            return p.val == q.val and self.isTreeSymmetric(p.left, q.right) and self.isTreeSymmetric(p.right, q.left)
        elif not p and not q:
            return True
        else:
            return False

File: 101-200/101.symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isTreeSymmetric(TreeNode* p, TreeNode* q) {
        if (p != nullptr && q != nullptr)
        {
            return p->val == q->val && isTreeSymmetric(p->left, q->right) && isTreeSymmetric(p->right, q->left);
        }
        else if (p == nullptr && q == nullptr)
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    bool isSymmetric(TreeNode* root) {
        if (root == nullptr)
        {
            return true;
        }

        return isTreeSymmetric(root->left, root->right);
    }
};